### PR TITLE
feat: stream terraform output on diff, deploy, and destroy

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
The goal is to 

- Remove the scrollback issues
- Make the CLI usable on smaller screeens
- Show an attribute level diff
- Reduce complexity in our CLI

### Preview

detailed-diff-3

### CODE REVIEW

File: packages/cdktf-cli/bin/cmds/ui/deploy.tsx [hunk 1]
```diff
@@ -97,159 +63,21 @@ export const Deploy = ({
...
-          </Box>
-        </Fragment>
-      );
-    }
-    default:
-      return (
-        <Text>{`An unknown state occured: ${lastProjectUpdate?.type}`}</Text>
-      );
-  }
+  return <StreamView logs={logEntries}>{bottomBar}</StreamView>;
```
Comment (ansgarm) [inline packages/cdktf-cli/bin/cmds/ui/deploy.tsx:82]: love this ❤️

File: packages/cdktf-cli/bin/cmds/ui/components/outputs.tsx [hunk 2]
```diff
@@ -13,17 +17,75 @@ function sanitize(value: any) {
...
+
+            if (isTerraformOutput(value[k1])) {
+              return -1;
+            }
+
+            if (isTerraformOutput(value[k2])) {
+              return 1;
+            }
+            return k1.localeCompare(k2);
+          })
```
Comment (ansgarm) [inline packages/cdktf-cli/bin/cmds/ui/components/outputs.tsx:69]: nit: maybe move this into a comparator function with a very explicit name? would make the code more readable.

File: test/typescript/providers/test.ts [hunk 3]
```diff
@@ -7,7 +7,9 @@ describe("full integration test", () => {
       // This is a workaround for that, by having a new driver every run we don't have this bug.
       const deployDriver = new TestDriver(__dirname);
       await deployDriver.setupTypescriptProject();
-      expect(deployDriver.deploy("using-all-providers")).toMatchSnapshot();
+      const deployLog = deployDriver.deploy("using-all-providers");
+      expect(deployLog).toContain("1 to add");
+      expect(deployLog).toContain("Apply complete!");
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/typescript/providers/test.ts:12]: The problem here is that the diff looks different in terraform 0.15 vs terraform 1.0. I battled alternately failing tests a bit until I understood :D

File: test/typescript/multiple-stacks/test.ts [hunk 4]
```diff
@@ -77,55 +69,17 @@ describe("multiple stacks", () => {
...
-              " Destroying Stack: second
-              Resources
-               ✔ NULL_RESOURCE       test                null_resource.test
-
-
-              Summary: 1 destroyed.
-              "
-          `);
+      expect(driver.destroy("first")).toContain(`Destroy complete!`);
+      expect(driver.destroy("second")).toContain(`Destroy complete!`);
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/typescript/multiple-stacks/test.ts:82]: The output looks different per terraform version now

File: packages/cdktf-cli/lib/index.ts [hunk 5]
```diff
@@ -20,22 +20,6 @@ import { logger } from "./logging";
 
 export { SynthesizedStack, Language };
 
-export enum Status {
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/lib/index.ts:23]: yes, not needed anymore

File: packages/cdktf-cli/lib/models/terraform-cli.ts [hunk 6]
```diff
@@ -75,7 +75,8 @@ export class TerraformCli implements Terraform {
       terraformBinaryName,
       ["show", "-json", planFile],
       { cwd: this.workdir, env: process.env },
-      this.onStdout("plan"),
+      // we don't care about the output, this is just internally to compose a plan
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/lib/models/terraform-cli.ts:78]: yes, the output was printed before, this added a json view of a planfile to the logs

File: packages/cdktf-cli/test/lib/index.test.ts [hunk 7]
```diff
@@ -116,10 +116,9 @@ describe("CdktfProject", () => {
         },
       });
 
-      return expect(cdktfProject.diff()).rejects.toMatchInlineSnapshot(`
-                "terraform init errored with: 
-                Error: Usage Error: Found more than one stack, please specify a target stack. Run cdktf diff <stack> with one of these stacks: first, second "
-              `);
+      return expect(cdktfProject.diff()).rejects.toMatchInlineSnapshot(
+        `[Error: terraform init errored with: Error: Usage Error: Found more than one stack, please specify a target stack. Run cdktf diff <stack> with one of these stacks: first, second ]`
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/test/lib/index.test.ts:120]: You are right, I will completely remove the prefix from the error

File: packages/cdktf-cli/lib/project-execution.ts [hunk 8]
```diff
@@ -284,7 +303,8 @@ const projectExecutionMachine = createMachine<ProjectContext, ProjectEvent>(
           onError: {
             target: "error",
             actions: assign({
-              message: (_context, event) => event.data,
+              message: (_context, event) =>
+                extractError("Synth errors")(_context, event),
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/lib/project-execution.ts:307]: I completely removed the prefixing, there is enough context around

File: packages/cdktf-cli/test/lib/index.test.ts [hunk 9]
```diff
@@ -116,10 +116,9 @@ describe("CdktfProject", () => {
         },
       });
 
-      return expect(cdktfProject.diff()).rejects.toMatchInlineSnapshot(`
-                "terraform init errored with: 
-                Error: Usage Error: Found more than one stack, please specify a target stack. Run cdktf diff <stack> with one of these stacks: first, second "
-              `);
+      return expect(cdktfProject.diff()).rejects.toMatchInlineSnapshot(
+        `[Error: Error: Usage Error: Found more than one stack, please specify a target stack. Run cdktf diff <stack> with one of these stacks: first, second ]`
```
Comment (ansgarm) [inline packages/cdktf-cli/test/lib/index.test.ts:120]: Is `Error: Error: Usage Error:` shown to users? 😅

**Review by ansgarm — APPROVED** [review]

File: packages/cdktf-cli/test/lib/index.test.ts (hunk 9 above)
Comment (DanielMSchmidt) [reply packages/cdktf-cli/test/lib/index.test.ts:120]: No, `Error: Usage Error:` is shown

**Pull request merged**